### ISSUE TITLE
Question: Path of least resistance to WinUI 3.0 + Win32?

### ISSUE THREAD

**Issue by dpaulino** [microsoft/microsoft-ui-xaml#1945]
Hi all, 

Let's say I wanted to build a new app, and my end goal is to make this app a Win32 desktop app using the latest .NET Core 3 (or .NET 5) but use WinUI for its front end. Obviously, I cannot do this right now (since WinUI 3.0 is still in alpha). So I'm wondering which of these approaches would lead to give me the fewest headaches:

1. Create a UWP app to start and then migrate this app to use WinUI 3.0 + Win32 (.NET Core 3 or .NET 5)
2. Create a WPF app to start, use XAML islands in UI, and then migrate this app to WinUI 3.0.

Thoughts? And sorry if I'm misusing some terminology in my question. I'm still wrapping my head around all this.

**Comment by robloo** [comment]
Just my two cents: I would go with path 1 and create a UWP app to start. However this answer is highly dependent on exactly what you intend to do.
 * It assumes you can do the vast majority of your backend code in managed C# using .NET Core 2.0 to start and you won't be doing much with Win32 although interop is possible. Win32 from UWP is usually a problem. UWP also doesn't support anything higher than .NET Core 2.0 at the moment (it's on life support to be replaced by WinUI).
 * Converting from UWP XAML to WinUI 3.0 XAML is only a change of namespace. This is not true of WPF XAML to WinUI 3.0 XAML. While converting from WPF to WinUI XAML is usually not difficult, there are several differences including: 
    * x:Uid for localization in UWP only (and much easier than WPF)
    * x:Bind in UWP only which is more performant and easier to debug. Definately recommended over Binding
    * x:Static and x:Type don't exist in UWP
     * OneWayToSource BindingMode is not in UWP
     * MultiBinding/IMultiBindingConverter doesn't exist in UWP -- this is a big one if you are doing a lot of MVVM work coming from a WPF background.
     * DataTriggers in Styling doesn't exist in UWP and does in WPF. VisualStateManager is now used in UWP but doens't exist in WPF. This is a non-trivial difference that requires some thought when converting styles and templates.
     * There is no LayoutTransform in UWP anymore

I could keep going but you get the idea. Starting with UWP means you don't have to worry about all of the above later as WinUI does it the same as UWP.

Bottom line though: unless you are developing with a large team, the code base isn't going to be so big in a year (about when WinUI 3.0 releases) that switching over either way becomes a huge problem. So will you need a lot of Win32 API's or not is seemingly the deciding factor.

**Comment by Felix-Dev** [comment]
@robloo
> VisualStateManager is now used in UWP but doens't exist in WPF. This is a non-trivial difference that requires some thought when converting styles and templates.

VisualStateManager does exist in WPF (it was added in .NET Framework 4.0). It's not as elegant as in UWP though, i.e. it has no [VisuaStateManager.Setters](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.visualstate.setters) property. That means you have to use Storyboards to set your values.

> MultiBinding/IMultiBindingConverter doesn't exist in UWP -- this is a big one if you are doing a lot of MVVM work coming from a WPF background.
This can be easily worked around with the new UWP feature of [function binding](https://docs.microsoft.com/en-us/windows/uwp/data-binding/function-bindings) (you can basically copy re-use your converter logic).

**Comment by Felix-Dev** [comment]
@dpaulino Ultimately, as @robloo said it, it depends on your app. No solution is perfect for now. With UWP you can write the UI once but you might to invest in workarounds you otherwise wouldn't need if you made a Win32/WPF app (i.e. call certain Win32 APIs which will require using an app service, that is a main UWP app with a helper full-trust win32 app). Then there's is also the sandbox you will have to accoutn for when designing the UX of your app.

If you decide to use XAML Islands this will limit you to 1903 and above currently. As WinUI Desktop is still quite a few months away from release, this could turn out to be an issue. There might also be additional development time required in getting your XAML Island and non XAML Island UI elements to work with each other (i.e. Binding)

If you decide to go the WPF route with MSIX only (so no XALM Island) but think you can at least start building logic using the Windows Runtime APIs, be aware that there might be unexpected roadblocks awaiting you (see issues https://github.com/microsoft/microsoft-ui-xaml/issues/1919 and https://github.com/microsoft/microsoft-ui-xaml/issues/1939 for example).

Frankly, I'm afraid each possible solution will require workarounds and it will depend on your actual app which framework/platform will ne the "path of least resistance".

Not to forget, we currently don't have a complete picture about WinUI 3 Desktop yet to say what will work and what will not when it will be released.

**Comment by Austin-Lamb** [comment]
Hi @dpaulino!

I would recommend starting with XAML Islands and WinUI 2.x.  That way all the UI you build, all the bindings you set up, all the localization you do with x:Uid, and so on, will all translate well to WinUI 3.  To do this, you'd have a host framework like WPF or Windows Forms that you use just to basically get the window on-screen, set a title on the window, stuff like that, and the entire window can be filled with one big XAML Island.

Then, when WinUI 3 comes out, you can upgrade to that and either use the "big island" approach (if you want to), or you can remove the hosting framework and use "WinUI for Desktop Apps" to have the entire window managed by the WinUI framework.  Your XAML markup, localization, bindings, and code-behind should basically just transfer over, except for some things like changing the "using namespaces" from Windows.UI.Xaml to Microsoft.UI.Xaml and stuff of that nature.

**Comment by dpaulino** [comment]
@Austin-Lamb I haven't had a chance to use xaml islands yet. Is it capable of running all uwp controls? Or does it only have support for a subset?

**Comment by Felix-Dev** [comment]
@dpaulino Since you host your custom UWP control (which can exist of multiple controls - think UserControl) in your WPF app that should be the case. At least I haven't noticed any UWP control not yet working. The [documentation](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/host-custom-control-with-xaml-islands) is also not mentioning any restrictions regarding UWP controls. If @Austin-Lamb knows of some controls not working and they are not mentioned in the docs - this info should be added then.

**Comment by dpaulino** [comment]
Thanks everyone for the responses. It sounds like both paths will cause me headaches. I've yet to make a decision but it sounds a little bit like waiting for WinUI 3.0 might have to be a third path to consider. I've gathered a lot of insight from this thread, so I'm comfortable closing it. Thanks again.
